### ISSUE TITLE
Change Request: Processor may provide parser options

### ISSUE THREAD

**Issue by DMartens** [eslint/eslint#18267]
### ESLint version

HEAD

### What problem do you want to solve?

Currently the processor API does not allow to provide parser options for the blocks returned from `preprocess`.
There are some cases where this is useful:
- `<script>` inside HTML may be `sourceType` script or module based on its type (the best currently option is to use `.mjs` file extension for the block for modules)
- non-module `<script>` inside HTML may declare globals
- markdown code blocks with language `node` shoulde have `sourceType: commonjs`

Currently the static properties such as `sourceType` may be handled via user configuration by using special document file names which is clunky.

### What do you think is the correct solution?

Allow returning `parserOptions` as part of a block in the `preprocess`, e.g. `return [{ text: '...', filename: '...', parserOptions: { ... } }]`.
The parserOptions would be preferred over user configuration.

### Participation

- [X] I am willing to submit a pull request for this change.

**Comment by nzakas** [comment]
Processors are agnostic as to how the code blocks they return will be parsed. Their only job is to take source code and return an array of code blocks. They should never be determining how those code blocks should be parsed.

People can manually configure any language options for code blocks in their configuration file:

```js
export default [
    { files: ["foo/bar.md/0.js"], languageOptions: { ecmaVersion: 5, sourceType: "module" } }
];
```

The code blocks are numbered, so they can target specific ones if necessary.

This will be easier if we implement https://github.com/eslint/eslint-plugin-markdown/pull/230
